### ISSUE TITLE
Expose Acorn allowImportExportEverywhere option

### ISSUE THREAD

**Issue by rhettlivingston** [eslint/eslint#6472]
Meteor version 1.3.3.1, just released, allows imports to be made from within conditionals and nested scopes. This was a welcome change.

Regrettably, it causes an actual parsing error in ESLint.
```
 272:7  error  Parsing error: 'import' and 'export' may only appear at the top level
```

I dissected enough of your code to find that Acorn has an option that adds the support we in the Meteor community need - the allowImportExportEverywhere option. I added code in my local copy of Espree to accept allowImportExportEverywhere as an ecmaFeature and pass it through to Acorn. It took all of three lines and accomplished exactly what I'd hoped.

In espree.js:
```js
        // apply parsing flags after sourceType to allow overriding
        if (options.ecmaFeatures && typeof options.ecmaFeatures === "object") {
            extra.ecmaFeatures = options.ecmaFeatures;
            impliedStrict = extra.ecmaFeatures.impliedStrict;
            extra.ecmaFeatures.impliedStrict = typeof impliedStrict === "boolean" && impliedStrict;
            if (options.ecmaFeatures.globalReturn) {
                acornOptions.allowReturnOutsideFunction = true;
            }
// begin new code
            if (options.ecmaFeatures.allowImportExportEverywhere) {
                acornOptions.allowImportExportEverywhere = true;
            }
// end new code
        }
```
In addition, it looks as though changing the current "meteor" environment implementation as follows makes it default that way for Meteor users.

In environments.js:
```js
    meteor: {
        globals: globals.meteor,
        parserOptions: {
          ecmaFeatures: {
            allowImportExportEverywhere: true
          }
        }
    },
```

I've read the notes at the bottom of your READMEs in both ESLint and Espree discouraging "experimental features". This one isn't experimental within the Meteor community now, it's production. And it promises to lead to exciting new capabilities in the future such as automatic code splitting. Furthermore, the NPM "reify" module that implements it is available to the whole NPM community, and Meteor itself will soon be moving to NPM. So it will spread beyond the Meteor community.

Please consider the above changes.

Also, I'd be happy to create PRs though I believe the above is basically it.

**Comment by ilyavolodin** [comment]
I don't know much about Meteor, but as far as I understand, this is incorrect implementation of imports. To the best of my knowledge ES6 spec doesn't allow none top-level imports (please correct me if I'm wrong). If that's the case, I don't think we can implement this change, since it would cause errors for users not using Meteor.
Btw, you can already get the same behavior if you use `babel-eslint`. As far as I understand, babel also implements imports incorrectly, and allows them anywhere.

**Comment by platinumazure** [comment]
@ilyavolodin Going against spec and wanting to maintain purity in that regard is one thing, but I don't understand your claim about causing errors for users not using Meteor. Users who don't want this behavior simply don't have to include this argument in ecmaFeatures. Am I missing something here?

**Comment by ilyavolodin** [comment]
@platinumazure No, you are not. However, if some third-party config, that you are extending enables this feature, your code will fail pretty much everywhere (browsers, node, etc.) You might not even be aware that the option is on.

**Comment by rhettlivingston** [comment]
First, I tried babel-eslint and it caused a parser error also.

I think that the idea that this would cause errors for those with mixed code environments (which Meteor is) is overstated. If the user accidentally has this turned on, the parser merely becomes **more** permissive, not less. So, the only error that it might cause is to not report an import as incorrectly placed below the top level in some code. I think that is a rare error that will likely become quickly obvious when they execute their code. Parse errors in environments that don't support this feature are not hard to find.

On the other hand, not having this causes ESLint to be useless in Meteor modules that contain nested imports because the parser error kills all other analysis of the module. And eslint-disable can't stop that.

Perhaps Meteor's tech lead, @benjamn, will chime in. He has also written a [defense of nested imports](https://github.com/benjamn/reify/blob/master/WHY_NEST_IMPORTS.md) that you may read.

The argument that convinces me to use them has to do with the fact that Meteor is an isomorphic environment that runs much of the same code on the server and client. We are now in the very weird position of having to use "require" in modules that are otherwise using import/export in order to conditionally include modules in a portion of a routine that needs to execute on the server environment only. So, we have two module systems within the same file. Ben's change eliminates that nastiness.

Lastly, Meteor is abandoning its package system and moving to NPM very soon. So, we in the Meteor community will be publishing our modules on NPM. Some will certainly have nested imports. This will spread.

**Comment by nzakas** [comment]
@platinumazure there is also the possibility that allowing this would cause our own rules to break. Everything is written with the assumption that `import` is a top-level node. We don't know what will break if that assumption is not true anymore. 

@rhettlivingston let's not turn this into a debate over whether or not nested imports are good, practical, or conventional. The only issue to discuss here is whether or not ESLint should support it out-of-the-box. 

In general, our position has always been that we support the official spec and anyone who wants something other than that will need to use a custom parser. Babel 5 supported nested imports, so perhaps using an older version of babel-eslint is the right choice.

**Comment by rhettlivingston** [comment]
@nzakas, I understand. I come from the Ada and Mil-Spec world and have taken such positions many times in the past.

I don't think that sticking an entire community that is becoming cutting edge (Meteor has just been moved all the way to ECMA 7 and both NPM 6 and the latest Mongo are on the way) to an old version of babel-eslint will work. I think we'd have to fork something before that.

I would note that acorn supports this today. I simply can't get to the option with the way that Espree is written. Only Espree has to change here. ESLint seems to leave the understanding of the options to Espree instead of rejecting unknown options up front. The ESLint change I proposed in the "meteor" env is just a nicety.

I believe there are other options in acorn that aren't exposed too. Perhaps a general way to pass options to acorn is what is really needed. Something as simple as passing on any unrecognized option might work.

I will dig further into babel-eslint and babel and see if there is a way to turn on an option there. If they had the feature, maybe they did the same thing as acorn and turned it into an option. And perhaps there is a way to get that option turned on without changes in babel-eslint.

I'm not sure whether there are other implications of switching to babel-eslint though. My brief look at it gave me the impression that it had caveats. Are our users as well served by that path as your default parser?

**Comment by nzakas** [comment]
@rhettlivingston Acorn supports a single flag that allows both `import` and `export` to be placed anywhere - you can't have just `import` anywhere. That opens up the possibility for a lot of confusion. 

> I don't think that sticking an entire community that is becoming cutting edge (Meteor has just been moved all the way to ECMA 7 and both NPM 6 and the latest Mongo are on the way) to an old version of babel-eslint will work. I think we'd have to fork something before that.

Unfortunately, when a library introduces custom syntax, you end up in this situation and can't expect all tools to work. 

Acorn only added the flag because Babel used to allow it. We also got requests to support it at that time, and our conclusion was the same: we only want to support the standards, otherwise we will get stuck needing to implement custom syntax for every new hot tool or library that comes along. 

We don't allow direct interaction with Acorn because it shouldn't matter that Espree is based on Acorn (it wasn't always).

**Comment by nzakas** [comment]
Sorry, hit the wrong button. 

I might be wrong (@hzoo can correct me), but isn't it possible to create a Babel plugin to support your extra syntax? If so, I'd think that babel-eslint could work with it.

**Comment by rhettlivingston** [comment]
> I might be wrong (@hzoo can correct me), but isn't it possible to create a Babel plugin to support your extra syntax? If so, I'd think that babel-eslint could work with it.

Thank you very much for this suggestion. I will dig into babel-eslint again and see if such a thing is possible. I know that Ben Newman plans a Babel plugin that includes his "reify" package that transpiles nested imports and exports. But for this purpose, I don't need the actual implementation that transpiles it all, I just need the parser to accept it. So, if a babel plugin can be specified without forking things and can turn on the flag for parsing, that would be a start.

> Acorn only added the flag because Babel used to allow it.

Purely in the interest of history (which I find fun), the issue records in Acorn indicate that they added the flag when they "corrected" their parser to not support it on December 8, 2014. Babylon, the parser used by babel-eslint, states in their credits "Heavily based on acorn and acorn-jsx, thanks to the awesome work of @RReverser and @marijnh." I went back to the Babylon repository view of December 7, 2014 and found that their README at that time was still titled "Acorn" in blazing letters at the top and their main parser was in "acorn.js". So Babylon, which began its life based on acorn before that date in history, inherited its "problem" with nested imports from acorn and "fixed" them after Acorn did by adopting Acorn's fix.

**Comment by benjamn** [comment]
Allowing the ESLint parser to accept a syntax without a parse error is hardly the same thing as supporting or encouraging that syntax. In fact, it would then be easier for ESLint to provide a meaningful lint warning/error about that syntax.

**Comment by hzoo** [comment]
We're passing this down to babylon in babel-eslint now (thanks to @rhettlivingston)

**Comment by rhettlivingston** [comment]
And thank you @hzoo for accepting the PR!

Though I've found no issues in babel-eslint's handling of Meteor code, I'd still like to see the same change in Espree at some point. I hope it doesn't have to wait until @benjamn or someone else can get the standard changed.

For the record, the option that babel-eslint now accepts is

```json
  "parserOptions": {
    "allowImportExportEverywhere": true
  },
```

**Comment by RReverser** [comment]
> I hope it doesn't have to wait until @benjamn or someone else can get the standard changed.

I guess misconception comes from here - the standard won't be changed to allow import/export everywhere exactly because it was meant to be used only on top-level and thus allow better static analysis than `require` can offer. It's kind of a syntactic guarantee that `import` / `export` won't be used in the runtime.

**Comment by benjamn** [comment]
@RReverser The top-level restriction is not the primary source of static analysis benefits: https://github.com/benjamn/reify/blob/master/WHY_NEST_IMPORTS.md#objections-and-critiques

We'll see what TC39 says next month.

**Comment by RReverser** [comment]
@benjamn My bad, didn't know that it's being discussed.

The reason I don't like that idea anymore (despite me being the one who defended it in Acorn) is that in e.g. browser context *loading* modules is relatively expensive operation, and when people see `import` inside of conditional, they might think that they're reducing traffic by not including polyfills etc., while in reality it will be loaded nevertheless, just not evaluated. In that sense, using explicit runtime function like `System.import` feels much more intuitive in what it does, when the module will be loaded and when executed.

Anyway, let's wait and see, as you said.

**Comment by nzakas** [comment]
Now supported via `babel-eslint` (thanks @hzoo!), so closing.
